client/verta/verta/runtime.py

```
@@ -0,0 +1,237 @@
+                         f"characters. (dashes and underscores permitted)")
+
+
+def log(key: str, value: Any) -> None:
+    """
+    Updates current logging context dict with provided key and value.
+    For use within a model's :meth:`~verta.registry.VertaModelBase.predict`
+    method to collect logging context.
+
+    .. note::
+        Multithreading of calls to log() within a model's predict() method is not
+        currently supported.
+
+    Parameters
+    ----------
+    key : str
+        String value to use as data label (key) in logging context, with a limit of
+        100 characters or less.
+    value : Any
+        Any `JSON serializable <https://docs.python.org/3/library/json.html#json.JSONEncoder>`__
+        value you wish to include in the context logs.
+    Returns
+    -------
+    None
+
+    Raises
+    -------
+    TypeError
+        If `validate` was set to ``True`` on the active :class:`context`, and
+        `value` is not a JSON-serializable type.
+    ValueError
+        If `key` provided contains non-alphanumeric characters.  Dashes `-`
+        and underscores `_` are permitted.
+
+    Examples
+    --------
+    .. code-block:: python
+       :emphasize-lines: 11,14
+
+        from verta import runtime
+
+        # Sample model code:
+        class MyModel(VertaModelBase):
+            def __init__(self, artifacts):
+                pass
+
+            @verify.io
+            def predict(self, data):
+                logs = {'some_stuff': 'I_care_about'}
+                runtime.log('my_logging_key', logs)
+                prediction_output = 'some_output'
+                more_logs = ['a', 'list', 'of', 'things']
+                runtime.log('things', more_logs)
+                return prediction_output
+
+    """
+    if _get_validate_flag():
+        _validate_json(value)
+    _validate_s3(key)
+    local_logs: Dict[str, Any] = _get_thread_logs()
+    local_logs.update({key: value})
```

I mentioned this in another comment, but we should raise an error if `log()` is called outside of a `context`.

---

Moving to a Slack thread for further discussion.